### PR TITLE
Add sai and esai initial driver support

### PR DESCRIPTION
**Pull request by dianacretu**
Add initial support for the i.MX8, i.MX8X and i.MX8M sai and esai driver.

### CODE REVIEW

File: include/hw/adsp/imx8m.h [hunk 1]
```diff
@@ -41,6 +41,9 @@
 #define ADSP_IMX8M_DSP_MU_BASE       XSHAL_MU2_SIDEB_BYPASS_PADDR
 #define ADSP_IMX8M_DSP_MU_SIZE         0x10000
 
+#define ADSP_IMX8M_ESAI_BASE       0x59010000
+#define ADSP_IMX8M_ESAI_SIZE       0x00010000
+
 #define ADSP_IMX8M_SAI_3_BASE     0x30c30000
```
Comment (dbaluta) [inline include/hw/adsp/imx8m.h:44]: @dianacretu there is no ESAI on i.MX8M as far as I know. So, lets remove this line.

File: hw/ssi/esai.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/* Virtualization support for ESAI.
+ *
+ * Copyright (C) 2020 NXP Corporation
+ *
+ * Author: Diana Cretu <diana.cretu@nxp.com>
```
Comment (dbaluta) [inline hw/ssi/esai.c:5]: Preferred way for NXP is: ```Copyright 2020 NXP```

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@dbaluta will merge when you are good.
@dianacretu the next stage is for this driver to read/write to a WAV file. We had this working on SSP a few years back but it's regressed with qemu updates so I will probably duplicate any code you have for sai etc into SSP, dmic etc.

**Comment by dbaluta** [comment]
@lgirdwood can you expand on this a little bit. We are still figuring our way with qemu.
At what point should we write to a wav file. 

at playback time:
- When there's a write to SAI FIFO TX register (usually by DMA).
at record time:
- When there's a read from SAI FIFO RX registers (again usually by DMA)

File: include/hw/ssi/esai.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/*
+ * BSD 3 Clause - See LICENCE file for details.
+ *
+ * Copyright (c) 2015, Intel Corporation
+ * All rights reserved.
+ */
```
Comment (dbaluta) [inline include/hw/ssi/esai.h:8]: Not sure about this. If you copied this from an Intel file we can still keep Intel license, but we also need to add

```Copyright 2020 NXP```

ESAI driver looks good to me. I have some concerns about SAI

File: hw/ssi/sai.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/* Virtualization support for SAI.
+ *
+ * Copyright 2020 NXP
+ *
```
Comment (dbaluta) [inline hw/ssi/sai.c:4]: We should be careful here. The instance of SAI integrated with imx8/imx8x indeed uses this register map. Anyhow, for imx8m the register map is a little bit different. 
See: https://github.com/thesofproject/sof/commit/182e62a198c3e700f6271ac986776f4ed8ace14a
Need to find a way to distinguish between when we compile this driver to be used with imx8 or imx8m

**Comment by lgirdwood** [comment]
> * When there's a write to SAI FIFO TX register (usually by DMA).
>   at record time:
> * When there's a read from SAI FIFO RX registers (again usually by DMA)

Yep, exactly - I did the same for SSP. When SSP enable bit was set (i.e. it was running).

e.g SSP capture. 
1) A qemu timer would be started on SSP enable bit.
2) timer work would load period size worth of data from WAV file into DMA buffer.
3) timer work would then assert a DMA IRQ.
4) DMA knows it has period worth of data and so on..

**Review by lgirdwood — COMMENTED** [review]
@dbaluta LGTM

File: include/hw/ssi/sai_imx8m.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*
+ * BSD 3 Clause - See LICENCE file for details.
```
Comment (dbaluta) [inline include/hw/ssi/sai_imx8m.h]: @dianacretu I think we can avoid having two sai headers file, by just adding an offset to sai.h

Like this;

```
#ifdef CONFIG_IMX8M
#define SAI_OFS		8
#else
#define SAI_OFS		0
#endif
```
Look at how SAI header is implemented in firmware code: https://github.com/thesofproject/sof/blob/182e62a198c3e700f6271ac986776f4ed8ace14a/src/include/sof/drivers/sai.h

**Comment by lgirdwood** [comment]
@dianacretu can you resolve the conflicts

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@dianacretu thanks. Looks good now.

**Pull request merged**